### PR TITLE
Add additional structured clone tests

### PR DESCRIPTION
**Pull request by andreubotella**
The following behaviors mandated by the spec weren't tested:

- Serializing a non-serializable platform object, or transferring a non-transferable platform object, should fail.
- Platform objects must deserialize even if their interface has been deleted from the realm's global.
- Objects which have serializable or transferable interfaces in their inheritance hierarchy must deserialize as instances of the closest such interface.
- Transferring again a detached object will fail.

This change also passes the test object as an additional parameter to the `f` function of structured clone tests, to make it possible to test that a promise rejects.

### CODE REVIEW

File: html/webappapis/structured-clone/structured-clone-battery-of-tests-with-transferables.js [hunk 1]
```diff
@@ -20,3 +20,67 @@ structuredCloneBatteryOfTests.push({
...
+      globalThis.MessagePort = messagePortInterface;
+    }
+  }
+});
+
+structuredCloneBatteryOfTests.push({
+  description: 'A subclass instance will be received as its closest transferable superclass',
+  async f(runner) {
+    // MessagePort doesn't have a constructor, so we must use something else.
+    class OffscreenCanvasSubclass extends OffscreenCanvas {}
```
Comment (domenic) [inline html/webappapis/structured-clone/structured-clone-battery-of-tests-with-transferables.js:81]: OffscreenCanvas isn't implemented in Firefox and Safari. Can we use something else? I think ImageBitmap would be good.

File: html/webappapis/structured-clone/structured-clone-battery-of-tests-with-transferables.js [hunk 2]
```diff
@@ -20,3 +20,67 @@ structuredCloneBatteryOfTests.push({
...
+    const original = new OffscreenCanvasSubclass(200, 200);
+    const transfer = await runner.structuredClone(original, [original]);
+    assert_equals(Object.getPrototypeOf(transfer), OffscreenCanvas.prototype);
+  }
+});
```
Comment (domenic) [inline html/webappapis/structured-clone/structured-clone-battery-of-tests-with-transferables.js:86]: Optional nit: newline at end of file would make me happier.

**Review by domenic — CHANGES_REQUESTED** [review]

File: html/webappapis/structured-clone/structured-clone-battery-of-tests-with-transferables.js (hunk 1 above)
Comment (andreubotella) [reply html/webappapis/structured-clone/structured-clone-battery-of-tests-with-transferables.js:81]: `ImageBitmap`, just like `MessagePort`, doesn't have a constructor, so you can't instantiate a subclass. You could create a new object that has `ImageBitmap.prototype` in its prototype chain, or you could change the prototype of an existing instance – which probably should be added as tests of their own – but neither is what I'm testing here.

The only constructable transferable interface that's implemented in every browser seems to be `ReadableStream`, and it's only transferable in Chrome.

Comment (domenic) [reply html/webappapis/structured-clone/structured-clone-battery-of-tests-with-transferables.js:81]: Well, hmm. I would slightly prefer ReadableStream if that still gets the same coverage, as transferability seems likely to be implemented sooner than OffscreenCanvas.

In either event, it sounds like this test should be made optional using https://web-platform-tests.org/writing-tests/testharness-api.html#optional-features

**Review by domenic — APPROVED** [review]

**Pull request merged**